crates/wasm-smith/src/component.rs

```diff
@@ -0,0 +1,802 @@
+            9 => Ok(CompoundType::Named(
+                self.arbitrary_named_type(u, &mut HashSet::new())?,
+            )),
+            _ => unreachable!(),
+        }
+    }
+
+    fn arbitrary_import_section(&mut self, u: &mut Unstructured<'a>) -> Result<()> {
+        todo!()
+    }
+
+    fn arbitrary_func_section(&mut self, u: &mut Unstructured<'a>) -> Result<()> {
+        todo!()
+    }
+
+    fn arbitrary_core_section(&mut self, u: &mut Unstructured<'a>) -> Result<()> {
+        todo!()
+    }
+
+    fn arbitrary_component_section(&mut self, u: &mut Unstructured<'a>) -> Result<()> {
+        todo!()
+    }
+
+    fn arbitrary_instance_section(&mut self, u: &mut Unstructured<'a>) -> Result<()> {
+        todo!()
+    }
+
+    fn arbitrary_export_section(&mut self, u: &mut Unstructured<'a>) -> Result<()> {
+        todo!()
+    }
+
+    fn arbitrary_start_section(&mut self, u: &mut Unstructured<'a>) -> Result<()> {
+        todo!()
+    }
+
+    fn arbitrary_alias_section(&mut self, u: &mut Unstructured<'a>) -> Result<()> {
+        todo!()
+    }
+}
+
+#[derive(Debug)]
+enum Section<'a> {
+    Custom(CustomSection<'a>),
+    Type(TypeSection),
+    Import(ImportSection),
+    Func(FuncSection),
+    Core(CoreSection),
+    Component(ComponentSection),
+    Instance(InstanceSection),
+    Export(ExportSection),
+    Start(StartSection),
+    Alias(AliasSection),
+}
+
+#[derive(Debug)]
+struct CustomSection<'a> {
+    name: &'a str,
+    data: &'a [u8],
+}
+
+impl<'a> Arbitrary<'a> for CustomSection<'a> {
+    fn arbitrary(u: &mut Unstructured<'a>) -> Result<Self> {
+        let name = limited_str(1_000, u)?;
+        let data = u.arbitrary()?;
+        Ok(CustomSection { name, data })
+    }
+}
+
+#[derive(Debug)]
+struct TypeSection {
+    types: Vec<Type>,
+}
+
+#[derive(Clone, Debug)]
+enum Type {
+    Module(ModuleType),
+    Component(ComponentType),
+    Instance(InstanceType),
+    Func(FuncType),
+    Value(ValueType),
+    Compound(CompoundType),
+}
+
+#[derive(Clone, Debug)]
+struct ModuleType {
+    defs: Vec<ModuleTypeDef>,
+}
+
+#[derive(Clone, Debug)]
+enum ModuleTypeDef {
+    TypeDef(crate::core::Type),
+    Import(crate::core::Import),
+    Export(String, crate::core::EntityType),
+}
+
+#[derive(Clone, Debug)]
+enum Alias {
+    InstanceExport {
+        instance: u32,
+        name: String,
+        kind: InstanceExportAliasKind,
+    },
+    Outer {
+        count: u32,
+        i: u32,
+        kind: OuterAliasKind,
+    },
+}
+
+#[derive(Clone, Debug)]
+enum InstanceExportAliasKind {
+    Module,
+    Component,
+    Instance,
+    Func,
+    Value,
+    Table,
+    Memory,
+    Global,
+}
+
+#[derive(Clone, Debug)]
+enum OuterAliasKind {
+    Module,
+    Component,
+    Type,
+}
+
+#[derive(Clone, Debug)]
+struct ComponentType {
+    defs: Vec<ComponentTypeDef>,
+}
+
+#[derive(Clone, Debug)]
+enum ComponentTypeDef {
+    Instance(InstanceTypeDef),
+    Import(Import),
+}
```

Comment:
Expanding on my comment from above, I thought that this would have an import, export, type, and alias case. I didn't think that instances would be singled-out here

Reply:
`InstanceTypeDef` has exports, types, and aliases. I could probably flatten it into this to make the types a bit more clear, but that would also mean duplicating the code to generate this stuff.

Reply:
I've flattened this now.